google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,6 +181,258 @@ class ResumableAsyncStreamingReadWriteRpc {
   virtual future<void> Shutdown() = 0;
 };
 
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      RetryPolicyFactory retry_factory,
+      std::shared_ptr<BackoffPolicy const> backoff_policy, AsyncSleeper sleeper,
+      AsyncStreamFactory<RequestType, ResponseType> stream_factory,
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_factory_(std::move(retry_factory)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ~ResumableAsyncStreamingReadWriteRpcImpl() override {
+    future<void> shutdown = Shutdown();
+    if (!shutdown.is_ready()) {
+      GCP_LOG(WARNING) << "`Finish` must be called and finished before object "
+                          "goes out of scope.";
+      assert(false);
+    }
+    shutdown.get();
+  }
+
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+      assert(!retry_promise_.has_value());
+      retry_promise_.emplace();
+    }
+    auto retry_policy = retry_factory_();
+    auto backoff_policy = backoff_policy_prototype_->clone();
+    Initialize(std::move(retry_policy), std::move(backoff_policy));
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    // TODO(18suresha) implement
+    return make_ready_future(absl::optional<ResponseType>());
+  }
+
+  future<bool> Write(RequestType const&, grpc::WriteOptions) override {
+    // TODO(18suresha) implement
+    return make_ready_future(false);
+  }
+
+  future<void> Shutdown() override {
+    promise<void> root_promise;
+    future<void> root_future = root_promise.get_future();
+    {
```

Comment:
Seems like it would be better to refactor all this code to a function so this reads:

```cc
future<void> Shutdown() override {
  promise<void> p;
  auto future = ScheduleShutdown( // <<< think of a better name
    p.get_future());
 p.set_value();
 return future;
}

...
...

future<void> ScheduleShutdown(future<void> root_future) {
  std::lock_guard<std::mutex> lk(mu_);
  // as before
}
```